bqskit/passes/mapping/layout/pam.py

```diff
@@ -82,10 +82,11 @@ async def run(self, circuit: Circuit, data: PassData) -> None:
         if not subgraph.is_fully_connected():
             raise RuntimeError('Cannot route circuit on disconnected qudits.')
 
-        pi = data.initial_mapping
+        pi = [i for i in range(circuit.num_qudits)]
```

Comment:
Yeah, good questions. A few points:

- The initial and final mappings are always with respect to the original circuit (mapping from original to current), so until the circuit gets modified (ApplyPlacement, Routing), they shouldn't be changed.
- The current placement tracks the current placement of the current `circuit` qudits to `model` qudits, so layout and routing need to start with this information
- pi always starts as [0, 1, 2, 3, ...], because the placement info is encoded into the `data.connectivity` call, the actual coupling graph is permuted there to account for it

Yes, ApplyPlacement should be called after SABRE layout and routing. What other workflows are you referring to?

Also, Rather than running `GreedyPlacement`, another valid placement strategy would be to `ApplyPlacement` before mapping to "extend" the circuit to the model size, then running layout and hoping SABRE finds a good placement. This is sort of the strategy we take with PAM. We are definitely lacking in the placement department tho

Reply:
Oh ok, so even before we were calling ApplyPlacement before and after PAMLayout and PAMRouting.